File: src/main/java/org/swordess/common/vitool/ext/AbstractShellComponent.java
package org.swordess.common.vitool.ext;

import org.springframework.shell.component.StringInput;
import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public class AbstractShellComponent extends org.springframework.shell.standard.AbstractShellComponent {

    public class Option {

        private final List<Supplier<String>> providers;

        private Option() {
            this(new ArrayList<>());
        }

        private Option(List<Supplier<String>> providers) {
            this.providers = providers;
        }

        Option or(Supplier<String> provider) {
            providers.add(provider);
            return this;
        }

        public Option orEnv(String envVarName) {
            return or(() -> System.getenv(envVarName));
        }

        public Option orInput(String prompt) {
            return orInput(prompt, null, true);
        }

        public Option orInput(String prompt, String defaultValue) {
            return orInput(prompt, defaultValue, true);
        }

        public Option orInput(String prompt, String defaultValue, boolean mask) {
            return or(() -> {
                StringInput component = new StringInput(getTerminal(), prompt, defaultValue);
                component.setResourceLoader(getResourceLoader());
                component.setTemplateExecutor(getTemplateExecutor());
                if (mask) {
                    component.setMaskCharater('*');
                }

                StringInput.StringInputContext context = component.run(StringInput.StringInputContext.empty());
                return context.getResultValue();
            });
        }

        public Option must(String errMsg) {
            return or(() -> {
                throw new IllegalArgumentException(errMsg);
            });
        }

        public String get() {
            for (Supplier<String> provider : providers) {
                String value = provider.get();
                if (StringUtils.hasText(value)) {
                    return value;
                }
            }
            return null;
        }

    }

    public Option optionValue(String value) {
        return new Option().or(() -> value);
    }

}

File: src/main/kotlin/org/swordess/common/vitool/ext/AbstractShellComponent.kt
/*
 * Copyright (c) 2019-2022 Swordess
 *
 * Distributed under MIT license.
 * See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
 */

package org.swordess.common.vitool.ext

import org.springframework.shell.component.StringInput
import org.springframework.shell.component.StringInput.StringInputContext
import org.springframework.shell.standard.AbstractShellComponent
import java.io.IOError
import java.io.InterruptedIOException
import java.lang.RuntimeException

open class AbstractShellComponent : AbstractShellComponent() {

    inner class Option(private val providers: MutableList<() -> String?> = mutableListOf()) {

        fun or(provider: () -> String?): Option {
            providers.add(provider)
            return this
        }

        fun orEnv(envVarName: String): Option = or { System.getenv(envVarName) }

        fun orInput(prompt: String, defaultValue: String? = null, mask: Boolean = true): Option = or {
            val component = StringInput(terminal, prompt, defaultValue)
            component.setResourceLoader(resourceLoader)
            component.templateExecutor = templateExecutor
            if (mask) {
                component.setMaskCharater('*')
            }

            try {
                val context = component.run(StringInputContext.empty())
                context.resultValue
            }catch (e: IOError) {
                if (e.cause is InterruptedIOException) {
                    throw RuntimeException("input has been cancelled")
                }
                throw e
            }
        }

        fun required(errMsg: String): Option = or { throw IllegalArgumentException(errMsg) }

        fun get(): String? {
            for (provider in providers) {
                val value = provider.invoke()
                if (value?.isNotBlank() == true) {
                    return value
                }
            }
            return null
        }

    }

    // String -> Option
    fun String?.orEnv(envVarName: String): Option = Option().or { this }.orEnv(envVarName)

    // Option -> String
    fun Option.must(errMsg: String): String = required(errMsg).get()!!

}